ably/realtime_presence.go

```diff
@@ -90,31 +119,72 @@ func (pres *RealtimePresence) send(msg *PresenceMessage) (result, error) {
 	}), nil
 }
 
-func (pres *RealtimePresence) syncWait() {
+func (pres *RealtimePresence) syncWait(ctx context.Context) error {
 	// If there's an undergoing sync operation or we wait till channel gets
 	// attached, the following lock is going to block until the operations
 	// complete.
-	pres.syncMtx.Lock()
-	pres.syncMtx.Unlock()
+	select {
+	case <-pres.syncDone:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// RTP18
+func syncSerial(msg *protocolMessage) (noChannelSerial bool, syncSequenceId string, syncCursor string) {
+	if empty(msg.ChannelSerial) { // RTP18c
+		noChannelSerial = true
+		return
+	}
+	// RTP18a
+	serials := strings.Split(msg.ChannelSerial, ":")
+	syncSequenceId = serials[0]
+	if len(serials) > 1 {
+		syncCursor = serials[1]
+	}
+	return false, syncSequenceId, syncCursor
 }
 
-func syncSerial(msg *protocolMessage) string {
-	if i := strings.IndexRune(msg.ChannelSerial, ':'); i != -1 {
-		return msg.ChannelSerial[i+1:]
+func (pres *RealtimePresence) enterMembers(internalMembers []*PresenceMessage) {
+	for _, member := range internalMembers {
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+		// RTP17g
+		err := pres.EnterClient(ctx, member.ClientID, member.Data)
+		// RTP17e
+		if err != nil {
+			pres.channel.log().Errorf("Error for internal member presence enter with id %v, clientId %v, err %v", member.ID, member.ClientID, err)
+			pres.channel.emitErrorUpdate(newError(91004, err), true)
+		}
+		cancel()
 	}
-	return ""
 }
 
-func (pres *RealtimePresence) onAttach(msg *protocolMessage) {
-	serial := syncSerial(msg)
+func (pres *RealtimePresence) onAttach(msg *protocolMessage, isAttachWithoutMessageLoss bool) {
```

Comment:
This is called from two places, both of which pass `true` as the second parameter. So can't we just get rid of that parameter and make the logic unconditional?

Reply:
updated